### PR TITLE
Use safe_malloc instead of new when creating new_envoy_map_entry

### PR DESCRIPTION
**Pull request by RyanTheOptimist**
Use safe_malloc instead of new when creating new_envoy_map_entry
since release_envoy_map uses free instead of delete. Found when 
building this code internally with some sort of strict checking enabled
    
Signed-off-by: Ryan Hamilton <rch@google.com>

Risk Level: Low
Testing: Existing tests
Docs Changes: N/A
Release Notes: N/A

### CODE REVIEW

**Comment by RyanTheOptimist** [comment]
@snowp I sent this your way based on the Obj-C/C/C++ discussion we had on Slack. Feel free to redirect though!

**Review by snowp — CHANGES_REQUESTED** [review]
Nice this makes sense to me, one suggestion

Is this check something we can enable for the EM build? Or is this some internal secret sauce?

File: test/common/http/header_utility_test.cc [hunk 1]
```diff
@@ -18,7 +18,8 @@ envoy_data envoyTestString(std::string& s, uint32_t* sentinel) {
 }
 
 TEST(RequestHeaderDataConstructorTest, FromCToCppEmpty) {
-  envoy_map_entry* header_array = new envoy_map_entry[0];
+  envoy_map_entry* header_array =
+      static_cast<envoy_map_entry*>(safe_malloc(sizeof(envoy_map_entry) * 0));
```
[the remark this answers is not in the document]
Comment (RyanTheOptimist) [reply test/common/http/header_utility_test.cc]: Good question! I was able to do this for both this test and the subsequent one. The rest of the tests, though, seem more elaborate. They set up envoy_map instances in which the envoy_data is wired up to use `envoy_test_release()` when releasing the data to do tracking. `makeEnvoyMap` seems to use `copyToBridgeData()` to populate the map which I think makes this a non-starter. But I'm not entirely familiar with this code so maybe there's a cleaner approach?

**Comment by RyanTheOptimist** [comment]
> Nice this makes sense to me, one suggestion
> 
> Is this check something we can enable for the EM build? Or is this some internal secret sauce?

Interesting question. It does look like an internal secret sauce thing in tcmalloc, according to some internal docs. However, the chromium version of tcmalloco has something basically similar. There error comes from here:

https://chromium.googlesource.com/chromium/src/third_party/tcmalloc/chromium/+/refs/heads/main/src/debugallocation.cc#474

So it might be something to investigate, but i don't think it works out of the box.

**Review by snowp — APPROVED** [review]
Thanks!

**Pull request merged**